src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -947,4 +982,10 @@ public String concreteIndex() {
             return concreteIndex;
         }
     }
+
+    void decrementCounter(@Nullable IndexShard indexShard) {
```

Comment:
I think we can simplify this a bit  and maybe have a class that does that ie this:

```Java

static class IndexShardReference {
  private final AtomicReference<IndexShard> ref = new AtomicReference();
  
  void setReference(IndexShard shard) {
     indexShard.incrementOperationCounter();
     assert ref.get() == null;
     ref.set(indexShard);
  } 

  void removeReference() {
    IndexShard shard = ref.get();
    try {
       shard. decrementOperationCounter();
    } finally {
      ref.set(null);
    } 
  }
}

```

this seems cleaner? or do I miss something